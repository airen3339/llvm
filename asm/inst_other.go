package asm

import (
	"fmt"

	"github.com/llir/llvm/ir"
	"github.com/llir/ll/ast"
)

// --- [ Other instructions ] --------------------------------------------------

// ~~~ [ icmp ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstICmp(inst ir.Instruction, old *ast.ICmpInst) (*ir.InstICmp, error) {
	i, ok := inst.(*ir.InstICmp)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstICmp, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ fcmp ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFCmp(inst ir.Instruction, old *ast.FCmpInst) (*ir.InstFCmp, error) {
	i, ok := inst.(*ir.InstFCmp)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFCmp, got %T", inst))
	}
	// Fast math flags.
	i.FastMathFlags = irFastMathFlags(old.FastMathFlags())
	// TODO: implement
	return i, nil
}

// ~~~ [ phi ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstPhi(inst ir.Instruction, old *ast.PhiInst) (*ir.InstPhi, error) {
	i, ok := inst.(*ir.InstPhi)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstPhi, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ select ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstSelect(inst ir.Instruction, old *ast.SelectInst) (*ir.InstSelect, error) {
	i, ok := inst.(*ir.InstSelect)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstSelect, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ call ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstCall(inst ir.Instruction, old *ast.CallInst) (*ir.InstCall, error) {
	i, ok := inst.(*ir.InstCall)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstCall, got %T", inst))
	}
	// Fast math flags.
	i.FastMathFlags = irFastMathFlags(old.FastMathFlags())
	// Calling convention.
	i.CallingConv = irOptCallingConv(old.CallingConv())
	return i, nil
}

// ~~~ [ va_arg ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstVAArg(inst ir.Instruction, old *ast.VAArgInst) (*ir.InstVAArg, error) {
	i, ok := inst.(*ir.InstVAArg)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstVAArg, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ landingpad ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstLandingPad(inst ir.Instruction, old *ast.LandingPadInst) (*ir.InstLandingPad, error) {
	i, ok := inst.(*ir.InstLandingPad)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstLandingPad, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ catchpad ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstCatchPad(inst ir.Instruction, old *ast.CatchPadInst) (*ir.InstCatchPad, error) {
	i, ok := inst.(*ir.InstCatchPad)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstCatchPad, got %T", inst))
	}
	// TODO: implement
	return i, nil
}

// ~~~ [ cleanuppad ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstCleanupPad(inst ir.Instruction, old *ast.CleanupPadInst) (*ir.InstCleanupPad, error) {
	i, ok := inst.(*ir.InstCleanupPad)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstCleanupPad, got %T", inst))
	}
	// TODO: implement
	return i, nil
}
