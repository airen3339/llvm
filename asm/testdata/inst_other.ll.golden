@g1 = global i32 42

define i1 @icmp_1() {
; <label>:0
	%result = icmp ne i32 42, 5
	ret i1 %result
}

define <2 x i1> @icmp_2() {
; <label>:0
	%result = icmp eq <2 x i32> <i32 42, i32 11>, <i32 42, i32 22>
	ret <2 x i1> %result
}

define void @icmp_3() {
; <label>:0
	%1 = icmp eq i32 10, 15
	%2 = icmp ne i32 10, 15
	%3 = icmp ugt i32 10, 15
	%4 = icmp uge i32 10, 15
	%5 = icmp ult i32 10, 15
	%6 = icmp ule i32 10, 15
	%7 = icmp sgt i32 10, 15
	%8 = icmp sge i32 10, 15
	%9 = icmp slt i32 10, 15
	%10 = icmp sle i32 10, 15
	ret void
}

define i1 @icmp_4() {
; <label>:0
	%result = icmp ne i32 42, 5, !baz !{!"qux"}, !foo !{!"bar"}
	ret i1 %result
}

define i1 @fcmp_1() {
; <label>:0
	%result = fcmp one double 42.0, 5.0
	ret i1 %result
}

define <2 x i1> @fcmp_2() {
; <label>:0
	%result = fcmp oeq <2 x double> <double 42.0, double 11.0>, <double 42.0, double 22.0>
	ret <2 x i1> %result
}

define void @fcmp_3() {
; <label>:0
	%1 = fcmp false double 10.0, 15.0
	%2 = fcmp oeq double 10.0, 15.0
	%3 = fcmp ogt double 10.0, 15.0
	%4 = fcmp oge double 10.0, 15.0
	%5 = fcmp olt double 10.0, 15.0
	%6 = fcmp ole double 10.0, 15.0
	%7 = fcmp one double 10.0, 15.0
	%8 = fcmp ord double 10.0, 15.0
	%9 = fcmp ueq double 10.0, 15.0
	%10 = fcmp ugt double 10.0, 15.0
	%11 = fcmp uge double 10.0, 15.0
	%12 = fcmp ult double 10.0, 15.0
	%13 = fcmp ule double 10.0, 15.0
	%14 = fcmp une double 10.0, 15.0
	%15 = fcmp uno double 10.0, 15.0
	%16 = fcmp true double 10.0, 15.0
	ret void
}

define i1 @fcmp_4() {
; <label>:0
	%result = fcmp one double 42.0, 5.0
	ret i1 %result
}

define i1 @fcmp_5() {
; <label>:0
	%result = fcmp one double 42.0, 5.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret i1 %result
}

define i1 @fcmp_6() {
; <label>:0
	%result = fcmp one double 42.0, 5.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret i1 %result
}

define i32 @phi_1(i1 %cond) {
entry:
	%result = phi i32 [ 42, %entry ]
	ret i32 %result
}

define i32 @phi_2(i1 %cond) {
; <label>:0
	br i1 %cond, label %foo, label %bar
foo:
	br label %baz
bar:
	br label %baz
baz:
	%result = phi i32 [ 42, %foo ], [ 37, %bar ]
	ret i32 %result
}

define void @phi_3(i1 %cond) {
; <label>:0
	br i1 %cond, label %foo, label %bar
foo:
	%x = fadd double 32.0, 10.0
	br label %baz
bar:
	br label %baz
baz:
	%1 = phi i32 [ 42, %foo ], [ 37, %bar ]
	%2 = phi i32* [ null, %foo ], [ @g1, %bar ]
	%3 = phi double [ %x, %foo ], [ 11.0, %bar ]
	%4 = phi { i32 } [ { i32 42 }, %foo ], [ zeroinitializer, %bar ]
	%5 = phi void ()* [ undef, %foo ], [ @j, %bar ]
	ret void
}

define i32 @phi_4(i1 %cond) {
; <label>:0
	br i1 %cond, label %foo, label %bar
foo:
	br label %baz
bar:
	br label %baz
baz:
	%result = phi i32 [ 42, %foo ], [ 37, %bar ], !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @select_1(i1 %cond) {
; <label>:0
	%result = select i1 %cond, i32 42, i32 37
	ret i32 %result
}

define <2 x i32> @select_2(i1 %cond) {
; <label>:0
	%result = select i1 %cond, <2 x i32> <i32 42, i32 37>, <2 x i32> <i32 11, i32 22>
	ret <2 x i32> %result
}

define i32 @select_3(i1 %cond) {
; <label>:0
	%result = select i1 %cond, i32 42, i32 37, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @f() {
; <label>:0
	ret i32 42
}

define i32 @call_1() {
; <label>:0
	%result = call i32 @f()
	ret i32 %result
}

define i32 @call_2() {
; <label>:0
	%result = call i32 @f()
	ret i32 %result
}

define i32 @call_3() {
; <label>:0
	%result = call i32 @f()
	ret i32 %result
}

define i32 @call_4() {
; <label>:0
	%result = call i32 @f()
	ret i32 %result
}

define double @g() {
; <label>:0
	ret double 42.0
}

define double @call_5() {
; <label>:0
	%result = call double @g()
	ret double %result
}

define void @call_6() {
; <label>:0
	%1 = call amdgpu_cs i32 @f()
	%2 = call amdgpu_gs i32 @f()
	%3 = call amdgpu_kernel i32 @f()
	%4 = call amdgpu_ps i32 @f()
	%5 = call amdgpu_vs i32 @f()
	%6 = call anyregcc i32 @f()
	%7 = call arm_aapcs_vfpcc i32 @f()
	%8 = call arm_aapcscc i32 @f()
	%9 = call arm_apcscc i32 @f()
	%10 = call avr_intrcc i32 @f()
	%11 = call avr_signalcc i32 @f()
	%12 = call ccc i32 @f()
	%13 = call fastcc i32 @f()
	%14 = call coldcc i32 @f()
	%15 = call ghccc i32 @f()
	%16 = call cc 11 i32 @f()
	%17 = call webkit_jscc i32 @f()
	%18 = call anyregcc i32 @f()
	%19 = call preserve_mostcc i32 @f()
	%20 = call preserve_allcc i32 @f()
	%21 = call swiftcc i32 @f()
	%22 = call cxx_fast_tlscc i32 @f()
	%23 = call x86_stdcallcc i32 @f()
	%24 = call x86_fastcallcc i32 @f()
	%25 = call arm_apcscc i32 @f()
	%26 = call arm_aapcscc i32 @f()
	%27 = call arm_aapcs_vfpcc i32 @f()
	%28 = call msp430_intrcc i32 @f()
	%29 = call x86_thiscallcc i32 @f()
	%30 = call ptx_kernel i32 @f()
	%31 = call ptx_device i32 @f()
	%32 = call spir_func i32 @f()
	%33 = call spir_kernel i32 @f()
	%34 = call intel_ocl_bicc i32 @f()
	%35 = call x86_64_sysvcc i32 @f()
	%36 = call x86_64_win64cc i32 @f()
	%37 = call x86_vectorcallcc i32 @f()
	%38 = call hhvmcc i32 @f()
	%39 = call hhvm_ccc i32 @f()
	%40 = call x86_intrcc i32 @f()
	%41 = call avr_intrcc i32 @f()
	%42 = call avr_signalcc i32 @f()
	%43 = call cc 86 i32 @f()
	%44 = call amdgpu_vs i32 @f()
	%45 = call amdgpu_gs i32 @f()
	%46 = call amdgpu_ps i32 @f()
	%47 = call amdgpu_cs i32 @f()
	%48 = call amdgpu_kernel i32 @f()
	%49 = call x86_regcallcc i32 @f()
	%50 = call ccc i32 @f()
	%51 = call coldcc i32 @f()
	%52 = call cxx_fast_tlscc i32 @f()
	%53 = call fastcc i32 @f()
	%54 = call ghccc i32 @f()
	%55 = call hhvm_ccc i32 @f()
	%56 = call hhvmcc i32 @f()
	%57 = call intel_ocl_bicc i32 @f()
	%58 = call msp430_intrcc i32 @f()
	%59 = call preserve_allcc i32 @f()
	%60 = call preserve_mostcc i32 @f()
	%61 = call ptx_device i32 @f()
	%62 = call ptx_kernel i32 @f()
	%63 = call spir_func i32 @f()
	%64 = call spir_kernel i32 @f()
	%65 = call swiftcc i32 @f()
	%66 = call webkit_jscc i32 @f()
	%67 = call x86_64_sysvcc i32 @f()
	%68 = call x86_64_win64cc i32 @f()
	%69 = call x86_fastcallcc i32 @f()
	%70 = call x86_intrcc i32 @f()
	%71 = call x86_regcallcc i32 @f()
	%72 = call x86_stdcallcc i32 @f()
	%73 = call x86_thiscallcc i32 @f()
	%74 = call x86_vectorcallcc i32 @f()
	ret void
}

define i32 ()* @h() {
; <label>:0
	ret i32 ()* @f
}

define void @call_7() {
; <label>:0
	%1 = call i32 @f()
	%2 = call i32 ()* @h()
	%3 = call i32 @f()
	%4 = call i32 @f()
	ret void
}

define i32 @i(i32 %x, ...) {
; <label>:0
	ret i32 42
}

define i32 @call_8() {
; <label>:0
	%result = call i32 (i32, ...) @i(i32 10, i32 20)
	ret i32 %result
}

define i32 @call_9() {
; <label>:0
	%result = call i32 @f()
	ret i32 %result
}

define i32 @call_10(i32 ()* %callee) {
; <label>:0
	%result = call i32 %callee()
	ret i32 %result
}

declare void @j()

define i32 @call_11() {
; <label>:0
	%result = call i32 bitcast (void ()* @j to i32 (i32)*)(i32 42)
	ret i32 %result
}

define i32 @call_12() {
; <label>:0
	%callee = bitcast void ()* @j to i32 (i32)*
	%result = call i32 %callee(i32 42)
	ret i32 %result
}

define i32 @call_13(i32 ()** %p) {
; <label>:0
	%callee = load i32 ()*, i32 ()** %p
	%result = call i32 %callee()
	ret i32 %result
}

define i32 @k(i32 %x, i32 %y) {
; <label>:0
	ret i32 42
}

define i32 @call_14() {
; <label>:0
	%result = call i32 @k(i32 11, i32 22)
	ret i32 %result
}

define i32 @call_15() {
; <label>:0
	%result = call i32 @f()
	ret i32 %result
}

define void @l() {
; <label>:0
	ret void
}

define void @call_16() {
; <label>:0
	call void @l()
	ret void
}

define i32 @call_17() {
; <label>:0
	%result = call i32 @f(), !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define double @m(double %x, double %y) {
; <label>:0
	ret double 42.0
}

define double @call_18() {
; <label>:0
	%result = call ccc double @m(double 11.0, double 22.0), !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}
